server/src/test/java/org/apache/druid/query/RetryQueryRunnerTest.java

```diff
@@ -0,0 +1,326 @@
+    final TimeseriesQuery query = timeseriesQuery(SCHEMA_INFO.getDataInterval());
+    final RetryQueryRunner<Result<TimeseriesResultValue>> queryRunner = createQueryRunner(
+        newRetryQueryRunnerConfig(1, true),
+        query
+    );
+    final Sequence<Result<TimeseriesResultValue>> sequence = queryRunner.run(QueryPlus.wrap(query));
+
+    // Let's move a segment
+    dropSegmentFromServerAndAddNewServerForSegment(servers.get(0));
+
+    final List<Result<TimeseriesResultValue>> queryResult = sequence.toList();
+    Assert.assertEquals(1, queryRunner.getNumTotalRetries());
+    // Note that we dropped a segment from a server, but it's still announced in the server view.
+    // As a result, we may get the full result or not depending on what server will get the retry query.
+    // If we hit the same server, the query will return incomplete result.
+    Assert.assertTrue(queryResult.size() > 8);
+    Assert.assertEquals(expectedTimeseriesResult(queryResult.size()), queryResult);
+  }
+
+  @Test
+  public void testRetryUntilWeGetFullResult()
+  {
+    prepareCluster(10);
+    final TimeseriesQuery query = timeseriesQuery(SCHEMA_INFO.getDataInterval());
+    final RetryQueryRunner<Result<TimeseriesResultValue>> queryRunner = createQueryRunner(
+        newRetryQueryRunnerConfig(100, false), // retry up to 100
+        query
+    );
+    final Sequence<Result<TimeseriesResultValue>> sequence = queryRunner.run(QueryPlus.wrap(query));
+
+    // Let's move a segment
+    dropSegmentFromServerAndAddNewServerForSegment(servers.get(0));
+
+    final List<Result<TimeseriesResultValue>> queryResult = sequence.toList();
+    Assert.assertTrue(0 < queryRunner.getNumTotalRetries());
+    Assert.assertEquals(expectedTimeseriesResult(10), queryResult);
+  }
+
+  @Test
+  public void testFailWithPartialResultsAfterRetry()
+  {
+    prepareCluster(10);
+    final TimeseriesQuery query = timeseriesQuery(SCHEMA_INFO.getDataInterval());
+    final RetryQueryRunner<Result<TimeseriesResultValue>> queryRunner = createQueryRunner(
+        newRetryQueryRunnerConfig(1, false),
+        query
+    );
+    final Sequence<Result<TimeseriesResultValue>> sequence = queryRunner.run(QueryPlus.wrap(query));
+    dropSegmentFromServer(servers.get(0));
+
+    expectedException.expect(SegmentMissingException.class);
+    expectedException.expectMessage("No results found for segments");
+    try {
+      sequence.toList();
+    }
+    finally {
+      Assert.assertEquals(1, queryRunner.getNumTotalRetries());
+    }
+  }
+
+  private void prepareCluster(int numServers)
+  {
+    for (int i = 0; i < numServers; i++) {
+      final DataSegment segment = newSegment(SCHEMA_INFO.getDataInterval(), i);
+      addServer(
+          SimpleServerView.createServer(i + 1),
+          segment,
+          segmentGenerator.generate(segment, SCHEMA_INFO, Granularities.NONE, 10)
+      );
+    }
+  }
+
+  private Pair<SegmentId, QueryableIndex> dropSegmentFromServer(DruidServer fromServer)
```

Comment:
Could you add some javadocs to these methods explaining what they do?

It's non-obvious that `dropSegmentFromServer` doesn't modify the server view, and that `dropSegmentFromServerAndAddNewServerForSegment` _does_ modify the server view, but only for the new server, not the old one.

Anywhere there is a lack of symmetry and obviousness like this, doc comments are especially important.

Reply:
Added for both methods.